### PR TITLE
MINOR: refactor FetchResponse#toMessage to avoid creating a lot of un…

### PR DESCRIPTION
**Pull request by chia7712**
Instead of creating intermediate collections to batch data, we directly check the last topic in auto-generated data to collect partitions having same topic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@ijuma This PR is from #9469 and it brings the better refactor :)

**Comment by ijuma** [comment]
Thanks for the PR. Have you checked that we get the same batching this way? The behavior won't be the same unless the iterator order is correct.

**Comment by chia7712** [comment]
@ijuma Thanks for your quick response!

> Have you checked that we get the same batching this way?

There is a test to verify the batch order (https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/server/FetchRequestTest.scala#L83).

>  The behavior won't be the same unless the iterator order is correct.

There are two rules of batching partition data.

1. the order must be kept. the tp order from input (iterator) must be equal to tp order in ```FetchResponseData```.
2. batch the partitions only if we can keep the their order in ```FetchResponseData```. For example: we should group t0-p0, t0-p1 and t1-p2 to two ```FetchableTopicResponse```s since the t0-p0 and t0-p1 can be merged. By contrast, t0-p0, t1-p1 and t0-p2 ought to generate three ```FetchableTopicResponse```s.

It seems to me the reason of generating the intermediate collections before is that reusing the data in ```Struct/Schema``` is not convenient.

File: clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java [hunk 1]
```diff
@@ -331,28 +331,32 @@ public int sessionId() {
...
-                partitionData.partitionResponse.setPartition(partitionId);
+        partIterator.forEachRemaining(entry -> {
+            PartitionData<T> partitionData = entry.getValue();
+            // Since PartitionData alone doesn't know the partition ID, we set it here
+            partitionData.partitionResponse.setPartition(entry.getKey().partition());
+            // We have to keep the order of input topic-partition. Hence, we batch the partitions only if the last
+            // batch is in the same topic group.
+            if (!topicResponseList.isEmpty() && topicResponseList.get(topicResponseList.size() - 1)
+                    .topic().equals(entry.getKey().topic())) {
+                topicResponseList.get(topicResponseList.size() - 1)
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java]: Maybe we could have a variable called `previousTopic` to make this code more readable.

**Review by ijuma — APPROVED** [review]
On closer inspection, I now remember that `batchByTopic` already relies on the ordering, so the new implementation is equivalent. LGTM overall. A nit below that should be easy to resolve.

File: clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java (hunk 1 above)
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java]: Nice suggestion. will address it later.

**Review by ijuma — APPROVED** [review]

**Pull request merged**